src/ConfigurationFactory.php

```diff
@@ -52,57 +52,61 @@ protected function createWithConfig(ContainerInterface $container, string $confi
-            $config['metadata_cache'],
-            'cache',
-            CacheFactory::class
-        );
-
-        $this->processCacheImplementation(
-            $configuration,
-            $metadataCache,
-            [$configuration, 'setMetadataCache']
-        );
-
-        $queryCache = $this->retrieveDependency(
-            $container,
-            $config['query_cache'],
-            'cache',
-            CacheFactory::class
-        );
-
-        $this->processCacheImplementation(
-            $configuration,
-            $queryCache,
-            [$configuration, 'setQueryCache']
-        );
-
-        $resultCache = $this->retrieveDependency(
-            $container,
-            $config['result_cache'],
-            'cache',
-            CacheFactory::class
-        );
-
-        $this->processCacheImplementation(
-            $configuration,
-            $resultCache,
-            [$configuration, 'setResultCache']
-        );
-
-        $hydrationCache = $this->retrieveDependency(
-            $container,
-            $config['hydration_cache'],
-            'cache',
-            CacheFactory::class
-        );
+        if (isset($config['metadata_cache'])) {
+            $metadataCache = $this->retrieveDependency(
+                $container,
+                $config['metadata_cache'],
+                'cache',
+                CacheFactory::class
+            );
+
+            $this->processCacheImplementation(
+                $metadataCache,
+                $configuration->setMetadataCache(...),
+            );
+        }
+
+        if (isset($config['query_cache'])) {
+            $queryCache = $this->retrieveDependency(
+                $container,
+                $config['query_cache'],
+                'cache',
+                CacheFactory::class
+            );
+
+            $this->processCacheImplementation(
+                $queryCache,
+                $configuration->setQueryCache(...),
+            );
+        }
+
+        if (isset($config['result_cache'])) {
+            $resultCache = $this->retrieveDependency(
+                $container,
+                $config['result_cache'],
+                'cache',
+                CacheFactory::class
+            );
 
-        $this->processCacheImplementation(
-            $configuration,
-            $hydrationCache,
-            [$configuration, 'setHydrationCache'],
-        );
+            $this->processCacheImplementation(
+                $resultCache,
+                $configuration->setResultCache(...),
+            );
+        }
+
+        if (isset($config['hydration_cache'])) {
+            $hydrationCache = $this->retrieveDependency(
+                $container,
+                $config['hydration_cache'],
+                'cache',
+                CacheFactory::class
+            );
+
+            $this->processCacheImplementation(
+                $hydrationCache,
+                $configuration->setHydrationCache(...),
+            );
+        }
```

Comment:
Would probably be best to **not** have these conditionals, and always fallback to a null/void cache for the basic setup?

Reply:
I believe that setter-injection is for optional dependencies, so I think no need to force clients to provide a default cache implementation

Reply:
Yes, but we can provide a default array or void cache for most of these, as stated below in https://github.com/Roave/psr-container-doctrine/pull/72#discussion_r1097132938

Reply:
🤔 It's a shame `psr/cache` v2 doesn't come with a null/void cache implementation since that's no doubt going to be the same for all implementations.

Reply:
ok, I will try add null cache implementation to activate optional cache and to simplify `ConfigurationFactory`. I do not think that `psr/cache` may get any implementation even if it useful in real world.